incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,90 @@
+package stats
+
+import (
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+func (c *counter) Incr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, 1)
+	*c = counter(i)
+}
+
+func (c *counter) Decr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, -1)
+	*c = counter(i)
+}
+
+type stats struct {
+	// HcTot is the total number of HierarchyConfig reconciliations.
+	HcTot counter
+
+	// HcCur is the currently undergoing number of HierarchyConfig reconciliations.
+	HcCur counter
+
+	// ObTot is the total number of object reconciliations.
+	ObTot counter
+
+	// ObCur is the currently undergoing number of object reconciliations.
+	ObCur counter
+
+	// HcRead is the total number of HierarchyConfig reads.
+	HcRead counter
+
+	// HcWrite is the total number of HierarchyConfig writes.
+	HcWrite counter
+
+	// ObRead is the total number of object reads.
+	ObRead counter
+
+	// ObWrite is the total number of object writes.
+	ObWrite counter
+}
+
+var Stats stats
```

Comment:
Not sure if we really need this level of grouping? They could all just be top-level variables, so you'd say `stats.ObjRead.Incr()` instead of `stats.Stats.ObjRead.Incr()` - the second `Stats` doesn't really add anything.

That's assuming you don't want to switch to a more functional style that I recommended above, like `stats.Objects.StartReconcile()` or even just `stats.StartObjectReconcile()`.

Reply:
I did more abstraction to have `stats.StartObjReconcile(gk)`, `stats.StartHierConfigReconcile()`, `stats.WriteHierConfig()`, etc. in the new change and kept `stats` to have `objects` field with `map[gk]*object` type.